### PR TITLE
Enable opt

### PR DESCRIPTION
**Pull request by ZoeLeibowitz**
The main new feature of this PR is the `PETScCallbackBuilder` class found inside `petsc/iet/routines.py`, which uses recursive compilation to generate each of the PETSc callback functions. 

Other small new features/changes:

- The `PETScSolve` in `petsc/solve.py` has been simplified due to the new recursive compilation approach. There is now just one equation (`InjectSolveEq`) that initially passes through the compilation lowering, serving as a placeholder for injecting calls to run the solver. All equations required by callback functions are attached to the `LinearSolveExpr` object, which forms the RHS of `InjectSolveEq`. These sets of equations are recursively compiled into the callback functions at the IET stage, as implemented within `PETScCallbackBuilder`. 

- `PETScCallable` subclass of `Callable` -> This allows optional parameters to appear in the function declaration even if the parameters are not used within the `Callable` .

- 2 new functions used within `_prepare_arguments` inside the `Operator` class -> This ensures that arguments passed to the `PETScStruct` (used to access context data in PETSc callback functions) are also "prepared" and can still be overridden with `.apply()`.

- Utility function `sort_petsc_frees` -> ensures PETSc objects are destroyed in a specified order.

- Some simplifications and edits to `lower_petsc` due to the new use of recursive compilation.

### CODE REVIEW

File: .github/workflows/examples-mpi.yml
Comment (ZoeLeibowitz) [inline .github/workflows/examples-mpi.yml:1]: These workflow edits are all temporary to trigger the CI..

File: devito/operator/operator.py [hunk 1]
```diff
@@ -509,15 +510,21 @@ def dimensions(self):
 
         # During compilation other Dimensions may have been produced
         dimensions = FindSymbols('dimensions').visit(self)
-        ret.update(d for d in dimensions if d.is_PerfKnob)
+
+        # NOTE: Should these dimensions be integrated into self._dimensions instead?
+        # In which case they would get picked up before this
+        struct_dims = derive_callback_dims(self._func_table)
```
Comment (FabioLuporini) [inline devito/operator/operator.py:516]: why shouldn't this logic be part of the FindSymbols visitor?

File: devito/passes/iet/definitions.py [hunk 2]
```diff
@@ -296,6 +297,7 @@ def _inject_definitions(self, iet, storage):
                 init = self.lang['thread-num'](retobj=tid)
                 frees.append(Block(header=header, body=[init] + body))
             frees.extend(as_list(cbody.frees) + flatten(v.frees))
+            frees = sort_petsc_frees(frees)
```
Comment (FabioLuporini) [inline devito/passes/iet/definitions.py:300]: petsc-specific actions within an agnostic pass looks bad

File: devito/operator/operator.py [hunk 3]
```diff
@@ -509,15 +510,21 @@ def dimensions(self):
...
+        ret.update(d for d in dimensions if d.is_PerfKnob or d in struct_dims)
 
         ret = tuple(sorted(ret, key=attrgetter('name')))
 
         return ret
 
     @cached_property
     def input(self):
-        return tuple(i for i in self.parameters if i.is_Input)
+        struct_params = derive_struct_inputs(self.parameters)
```
Comment (FabioLuporini) [inline devito/operator/operator.py:526]: I'm not sure I appreciate what's going on here

File: devito/petsc/iet/nodes.py [hunk 4]
```diff
@@ -1,31 +1,78 @@
...
+                 prefix=None, unused_parameters=None):
+        super().__init__(name, body, retval, parameters, prefix)
+        self._unused_parameters = as_tuple(unused_parameters)
+
+    @property
+    def unused_parameters(self):
+        return self._unused_parameters
+
+
+class Callback(Call):
```
Comment (FabioLuporini) [inline devito/petsc/iet/nodes.py:33]: why did you move this here? it's a generic node type, so it belongs to ir/iet/nodes.py

File: devito/petsc/types.py [hunk 5]
```diff
@@ -65,6 +94,12 @@ class SNES(LocalObject):
     """
     dtype = CustomDtype('SNES')
 
+    @property
+    def _C_free(self):
+        from devito.petsc.utils import petsc_call
```
Comment (FabioLuporini) [inline devito/petsc/types.py:99]: since you're in control of what happens within the `petsc/` module, this nested import shouldn't be to challenging to avoid?

File: devito/petsc/types.py [hunk 6]
```diff
@@ -28,21 +28,50 @@ def __init__(self, *args, stencil_width=None, **kwargs):
...
+class GlobalVec(LocalObject):
+    """
+    PETSc Vector object (Vec).
+    """
+    dtype = CustomDtype('Vec')
+
+    @property
+    def _C_free(self):
+        from devito.petsc.utils import petsc_call
+        destroy = petsc_call('VecDestroy', [Byref(self.function)])
```
Comment (FabioLuporini) [inline devito/petsc/types.py:71]: useless assignment

typically we keep the apparently useless assignments if the routine is relatively large , e.g. to easily inspect the variable content with a debugger, or because the assigned variable name encodes information about the variable content, but here it's not very useful

File: tests/test_iet.py [hunk 7]
```diff
@@ -129,26 +128,6 @@ def test_find_symbols_nested(mode, expected):
     assert [f.name for f in found] == eval(expected)
 
 
-def test_callback_cgen():
```
Comment (FabioLuporini) [inline tests/test_iet.py:132]: this should go back here

**Review by FabioLuporini — CHANGES_REQUESTED** [review]
My main objections here are the changes to

* operator.py
* definitions.py

Which we should try to avoid at costs. That's why I need to know what is exactly going on -- if petsc really is doing something so specific that we've never done until now or what

The rest looks great to me

File: devito/ir/iet/visitors.py [hunk 8]
```diff
@@ -314,7 +314,10 @@ def _args_call(self, args):
         return ret
 
     def _gen_signature(self, o, is_declaration=False):
-        decls = self._args_decl(o.parameters)
+        try:
+            decls = self._args_decl(o.parameters + o.unused_parameters)
```
Comment (FabioLuporini) [inline devito/ir/iet/visitors.py:318]: yet another petsc-specific edit, which we should somehow avoid

File: devito/petsc/iet/nodes.py [hunk 9]
```diff
@@ -1,31 +1,78 @@
...
+    def __init__(self, expr, pragmas=None, operation=OpInjectSolve):
         super().__init__(expr, pragmas=pragmas, operation=operation)
 
 
-class RHSLinearSystem(LinearSolverExpression):
+class PETScCallable(Callable):
+    def __init__(self, name, body, retval=None, parameters=None,
+                 prefix=None, unused_parameters=None):
+        super().__init__(name, body, retval, parameters, prefix)
+        self._unused_parameters = as_tuple(unused_parameters)
```
Comment (FabioLuporini) [inline devito/petsc/iet/nodes.py:26]: remind me what's this for?

Comment (mloubout) [reply devito/petsc/iet/nodes.py:26]: Where are those used by themselves? I can only find `paramerters+unused_parameters` to be used so not sure why it's needed separately.

File: devito/ir/iet/visitors.py [hunk 10]
```diff
@@ -672,6 +669,7 @@ def _operator_typedecls(self, o, mode='all'):
                 continue
             typedecls.extend([self._gen_struct_decl(j) for j in i.root.parameters
                               if xfilter(j)])
+            # from IPython import embed; embed()
```
Comment (EdCaunt) [inline devito/ir/iet/visitors.py]: Leftover

File: devito/petsc/iet/passes.py [hunk 11]
```diff
@@ -1,102 +1,100 @@
...
 
-        # Generate solver setup for target
-        for iter, (matvec,) in matvec_mapper.items():
-            # Skip the MatVecAction if it is not associated with the target
-            # There will most likely be more than one MatVecAction
-            # associated with the target e.g interior matvec + BC matvecs
-            if matvec.expr.rhs.target != target:
+        # Generate the solver setup for target. This is required only
+        # once per target
+        for _, (injectsolve,) in injectsolve_mapper.items():
```
[the remark this answers is not in the document]
Comment (ZoeLeibowitz) [reply devito/petsc/iet/passes.py]: Each `InjectSolveDummy` has a unique spatial iteration loop, which is then replaced by the calls to run the iterative solver

File: devito/petsc/iet/passes.py [hunk 12]
```diff
@@ -1,102 +1,100 @@
...
+            # Skip if not associated with the target
+            if injectsolve.expr.rhs.target != target:
                 continue
-            solver = generate_solver_calls(solver_objs, objs, matvec, target)
-            setup.extend(solver)
+            solver_setup = generate_solver_setup(solver_objs, objs, injectsolve, target)
+            setup.extend(solver_setup)
```
[the remark this answers is not in the document]
Comment (ZoeLeibowitz) [reply devito/petsc/iet/passes.py:67]: Based on our other conversation, I'm not going to use dictionaries because they don't integrate as cleanly with the second loop

File: devito/passes/iet/engine.py [hunk 13]
```diff
@@ -601,12 +601,12 @@ def update_args(root, efuncs, dag):
...
         foo(..., z) : root(x, z)
     """
-    if isinstance(root, ThreadCallable):
+    if isinstance(root, FixedArgsCallable):
         return efuncs
 
     # The parameters/arguments lists may have changed since a pass may have:
     # 1) introduced a new symbol
-    new_params = derive_parameters(root)
+    new_params = derive_parameters(root, True)
```
Comment (EdCaunt) [inline devito/passes/iet/engine.py:609]: Whatever arg `True` is passed to should probably be a kwarg

File: devito/petsc/iet/passes.py [hunk 14]
```diff
@@ -1,102 +1,99 @@
...
+    setup.extend([BlankLine, call_struct_callback] + calls_set_app_ctx)
 
-    # Remove the LinSolveExpr's from iet and efuncs
-    subs.update(rebuild_expr_mapper(iet))
     iet = Transformer(subs).visit(iet)
-    efuncs = transform_efuncs(efuncs, objs['struct'])
 
-    body = iet.body._rebuild(init=init, body=core+tuple(setup)+iet.body.body)
+    body = iet.body._rebuild(
+        init=init, body=core+tuple(setup)+(BlankLine,)+iet.body.body,
```
Comment (EdCaunt) [inline devito/petsc/iet/passes.py:91]: Body kwarg wants defining on a separate line imo

File: devito/petsc/iet/passes.py [hunk 15]
```diff
@@ -1,102 +1,99 @@
...
+        frees=(c.Line("PetscCall(PetscFinalize());"),)
+    )
     iet = iet._rebuild(body=body)
-
     metadata = core_metadata()
-    metadata.update({'efuncs': tuple(efuncs.values())})
+    metadata.update({'efuncs': tuple(builder.efuncs.values())+(struct_callback,)})
```
Comment (EdCaunt) [inline devito/petsc/iet/passes.py:96]: This tuple wants building on a separate line

File: devito/petsc/iet/routines.py [hunk 16]
```diff
@@ -0,0 +1,416 @@
...
+            dm_get_local_info
+        )
+
+        # Dereference function data in struct
+        dereference_funcs = [Dereference(i, struct) for i in
+                             struct.fields if isinstance(i.function, AbstractFunction)]
+
+        matvec_body = CallableBody(
+            List(body=body),
+            init=tuple([petsc_func_begin_user]),
```
Comment (EdCaunt) [inline devito/petsc/iet/routines.py:170]: `(petsc_func_begin_user,)`

File: devito/petsc/iet/routines.py [hunk 17]
```diff
@@ -0,0 +1,416 @@
...
+            stacks=stacks+tuple(dereference_funcs),
+            retstmt=tuple([Call('PetscFunctionReturn', arguments=[0])]))
```
Comment (EdCaunt) [inline devito/petsc/iet/routines.py:172]: (Call('PetscFunctionReturn', arguments=[0]),)

File: devito/petsc/iet/routines.py [hunk 18]
```diff
@@ -0,0 +1,416 @@
...
+            dm_get_local_info
+        )
+
+        # Dereference function data in struct
+        dereference_funcs = [Dereference(i, struct) for i in
+                             struct.fields if isinstance(i.function, AbstractFunction)]
+
+        formfunc_body = CallableBody(
+            List(body=body),
+            init=tuple([petsc_func_begin_user]),
```
Comment (EdCaunt) [inline devito/petsc/iet/routines.py:282]: Ditto

File: devito/petsc/iet/routines.py [hunk 19]
```diff
@@ -0,0 +1,416 @@
...
+            stacks=stacks+tuple(dereference_funcs),
+            retstmt=tuple([Call('PetscFunctionReturn', arguments=[0])]))
```
Comment (EdCaunt) [inline devito/petsc/iet/routines.py:284]: Same again

File: devito/petsc/iet/routines.py [hunk 20]
```diff
@@ -0,0 +1,416 @@
...
+            dm_get_local_info
+        )
+
+        # Dereference function data in struct
+        dereference_funcs = [Dereference(i, struct) for i in
+                             struct.fields if isinstance(i.function, AbstractFunction)]
+
+        formrhs_body = CallableBody(
+            List(body=[body]),
+            init=tuple([petsc_func_begin_user]),
```
Comment (EdCaunt) [inline devito/petsc/iet/routines.py:354]: +1

File: devito/petsc/iet/routines.py [hunk 21]
```diff
@@ -0,0 +1,416 @@
...
+            stacks=stacks+tuple(dereference_funcs),
+            retstmt=tuple([Call('PetscFunctionReturn', arguments=[0])]))
```
Comment (EdCaunt) [inline devito/petsc/iet/routines.py:356]: +1

**Review by EdCaunt — COMMENTED** [review]
Few minor comments but otherwise lgtm. @FabioLuporini @mloubout do you want to give this another review so Zoe can get this over the line and reduce the number of branches she's maintaining?

File: devito/ir/iet/nodes.py [hunk 22]
```diff
@@ -1048,7 +1057,12 @@ def functions(self):
 
     @property
     def expr_symbols(self):
-        ret = [self.pointer.indexed]
+        # TODO: Erick made a similar edit for NoDeclStruct so once merged,
+        # adjust accordingly
+        if self.pointer.is_Object:
```
Comment (FabioLuporini) [inline devito/ir/iet/nodes.py:1062]: this is not very neat

File: devito/operator/operator.py [hunk 23]
```diff
@@ -493,6 +493,9 @@ def _lower_iet(cls, uiet, profiler=None, **kwargs):
         # Target-independent optimizations
         minimize_symbols(graph)
 
+        # If necessary, sort frees into a specific order
+        sort_frees(graph)
```
Comment (FabioLuporini) [inline devito/operator/operator.py:497]: Imho this should enforced at construction time (inside `definitions.py`?), not through postprocessing like here

File: devito/types/array.py [hunk 24]
```diff
@@ -6,23 +6,25 @@
...
-from devito.types.basic import AbstractFunction
+from devito.types.basic import AbstractFunction, LocalType
 from devito.types.utils import CtypesFactory, DimensionTuple
 
 __all__ = ['Array', 'ArrayMapped', 'ArrayObject', 'PointerArray', 'Bundle',
            'ComponentAccess', 'Bag']
 
 
-class ArrayBasic(AbstractFunction):
+class ArrayBasic(AbstractFunction, LocalType):
```
Comment (FabioLuporini) [inline devito/types/array.py:16]: isn't this in devito master already? why does it appear here? do you still have to rebase?

File: devito/types/basic.py [hunk 25]
```diff
@@ -1443,6 +1444,37 @@ def __getnewargs_ex__(self):
         return args, kwargs
 
 
+class LocalType(Basic):
```
Comment (FabioLuporini) [inline devito/types/basic.py:1447]: same as above

**Review by FabioLuporini — COMMENTED** [review]
Aside from the comments in `ir/iet/nodes.py::Dereference::expr_symbols` and `sort_frees`, this big PR seems to be already in a very good state

File: devito/operator/operator.py (hunk 23 above)
Comment (ZoeLeibowitz) [reply devito/operator/operator.py]: I've moved the sort to `definitions.py`. It uses the new `_C_free_priority` property added to `LocalObject` types, as we briefly discussed...what do you think?...

**Review by FabioLuporini — APPROVED** [review]
